// mevn-cli/lib/utils/spinner.js
"use strict";
'user strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _ora = _interopRequireDefault(require("ora"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var Spinner = /*#__PURE__*/function () {
  function Spinner(text) {
    _classCallCheck(this, Spinner);

    // private member
    this.text = text;
  }

  _createClass(Spinner, [{
    key: "start",
    value: function start() {
      this.spinner = (0, _ora["default"])(this.text).start();
    }
  }, {
    key: "info",
    value: function info(text) {
      this.spinner.info(text);
    }
  }, {
    key: "succeed",
    value: function succeed(text) {
      this.spinner.succeed(text);
    }
  }, {
    key: "fail",
    value: function fail(text) {
      this.spinner.fail(text);
    }
  }, {
    key: "stop",
    value: function stop() {
      this.spinner.stop();
    }
  }]);

  return Spinner;
}();

var _default = Spinner;
exports["default"] = _default;
